### DISCUSSION TITLE
Check for empty workspaces

### DISCUSSION THREAD

**Discussion by trueNAHO** [leftwm/leftwm#1099] [Q&A]
How to [notify](https://github.com/leftwm/leftwm/blob/3fcd9e8df104000846ad09bb54f48ecf1299b851/themes/basic_eww/eww-bar/eww.yuck#L70%23L72) [eww](https://github.com/elkowar/eww) when a workspace focusses an empty tag without any windows?

The most related information from [`leftwm-state`](https://github.com/leftwm/leftwm/wiki/Customizing-Themes-with-%60liquid%60-templates#what-are-these-liquid-files-and-how-do-they-work) is:

```
 └ "workspaces"     <array of workspaces>
      └ "tags"        <array of tags for this workspace>
          ├ "index"     <index as int>
          ├ "mine"      <boolean is true, when the tag is owned by the active display>
          ├ "visible"   <boolean is true, when the tag active or is visible on another display>
          ├ "focused"   <boolean is true, when the focused window is on this tag>
          └ "busy"      <boolean is true, when the tag is not empty/has at least one window>
```

Tags with windows have their `busy` boolean set to `true`, while tags without windows have their `busy` boolean set to `false`. However, in `leftwm` version `0.4.2`, if a tag has the `mine` boolean set to `true`, the `busy` boolean is always set to `true`, regardless of whether a window exists on that tag.

**Comment by VuiMuich** [comment]
That should not be the case, and I am quite positive that on my system it is working correct.

I've setup eww to replace the tag string with `•` for empty tags, so I would have noticed immediately when all tags report `busy`.

Is it possible that for some reason you have a window like the bar, or the wallpaper set as `normal` window type? That would cause normally empty tags to register as `busy`.

**Reply by trueNAHO** [reply]
> Is it possible that for some reason you have a window like the bar, or the wallpaper set as `normal` window type? That would cause normally empty tags to register as `busy`.

My `~/.xinitrc` uses [feh](https://wiki.archlinux.org/title/Feh) for setting wallpapers. However, even after disabling everything in `~/.xinitrc` and removing configurations unrelated to `eww` in my [up](https://github.com/leftwm/leftwm/blob/3fcd9e8df104000846ad09bb54f48ecf1299b851/themes/basic_eww/up) script, the issue persisted.

To investigate further, I created a new user on my machine and tested the following Minimal Working Example (MWE):

# Minimal Working Example (MWE)

## `$HOME/.config/leftwm/themes/current`

This directory is a symbolic link to the [`basic_eww`](https://github.com/leftwm/leftwm/tree/3fcd9e8df104000846ad09bb54f48ecf1299b851/themes/basic_eww) theme.

## `$HOME/.config/eww/eww.yuck`

Use the following `$HOME/.config/eww/eww.yuck` file instead of the [`$HOME/.config/leftwm/themes/current/eww-bar`](https://github.com/leftwm/leftwm/tree/3fcd9e8df104000846ad09bb54f48ecf1299b851/themes/basic_eww/eww-bar) directory:

```lisp
(deflisten
 wmstate
 :initial '{"workspaces":[{"layout":"","tags":[{"name":"","index":0,"mine":false,"busy":false,"visible":false,"urgent":false}]}]}'
 "leftwm state")

(defwindow bar0
 :geometry (geometry :height "10px" :width "100%" :x "0" :y "0")
 :monitor 0
 :reserve (struts :distance "4%" :side "top")
 :stacking "fg"
 :windowtype "dock"
 (box
  (for tag in "${wmstate.workspaces[0].tags}"
   (button
    :onclick "leftwm-command 'SendWorkspaceToTag 0 ${tag.index}'"
    {tag.busy ? "*" : "${tag.name}"}))))
```

## Expected Behavior

Tags without any windows are denoted with `*`, and all other tags are denoted by their tag index.

## Actual Behavior

Only the selected tag is denoted with `*`, while all other tags are denoted by their tag index. For example, when no windows are open at all, cycling through all the tags results in the following sequence of tag names:

1. `{"*", "2", "3", "4", "5", "6", "7", "8", "9"}`
1. `{"1", "*", "3", "4", "5", "6", "7", "8", "9"}`
1. `{"1", "2", "*", "4", "5", "6", "7", "8", "9"}`
1. `{"1", "2", "3", "*", "5", "6", "7", "8", "9"}`
1. `{"1", "2", "3", "4", "*", "6", "7", "8", "9"}`
1. `{"1", "2", "3", "4", "5", "*", "7", "8", "9"}`
1. `{"1", "2", "3", "4", "5", "6", "*", "8", "9"}`
1. `{"1", "2", "3", "4", "5", "6", "7", "*", "9"}`
1. `{"1", "2", "3", "4", "5", "6", "7", "8", "*"}`

# Considerations

## Broken Setup

Please note that it is possible the issue might originate from an unrelated aspect of my broken root and user setup, which makes it difficult to track down. If we cannot resolve this issue, I will revisit it once my setup is no longer broken.

## New User

To minimize the risk of a broken setup, I have tested this MWE in a newly created user on my machine.

**Reply by VuiMuich** [reply]
:thinking:
This is my normal `box` for the tag widget:
```scheme
      (box
        (for tag in '${wmstate.workspaces[ws-id].tags}'
          (button
            :class {tag.mine ? "ws-button-mine" :
                    tag.visible ? "ws-button-visible" :
                    tag.urgent ? "ws-button-urgent" :
                    tag.busy ? "ws-button-busy" : "ws-button"}
            :onclick "leftwm-command \"SendWorkspaceToTag ${ws-id} ${tag.index}\""
            {!tag.mine && !tag.busy && !tag.visible && !tag.urgent ?  "·" : "${tag.name}"})))
```

I tried your example and negated `!tag.busy`, now only tags with windows and the visible tag show the tag string, the rest show `*`:
![2023-06-24-090032_512x20_scrot](https://github.com/leftwm/leftwm/assets/4779365/15444a60-b6a9-40ca-b969-1d2ca91323dc)

**Reply by nemalex** [reply]
I also noticed this behavior. The example works, but with `leftwm-state` there is no difference between a focused tag with 0 windows, and a focused tag with 0< windows.

**Reply by VuiMuich** [reply]
Ooh, I guess I misread the OP 🤦🏼‍♂️

In a focused Tag with now windows the `windows` array should be `[]`, no?

Currently on mobile so can't check rn.

**Reply by nemalex** [reply]
What windows array do you mean? The state doesn't have a `windows` field.

**Reply by trueNAHO** [reply]
Thank you, everyone, for your valuable input!

> In a focused Tag with now windows the `windows` array should be `[]`

Unfortunately, there seems to be no `windows` field. The [closest](https://github.com/leftwm/leftwm/wiki/Customizing-Themes-with-%60liquid%60-templates#what-are-these-liquid-files-and-how-do-they-work) one in behaviour is:

```
├ "window_title"   <window title as string>
```

> with `leftwm-state` there is no difference between a focused tag with 0 windows, and a focused tag with 0< windows

When no window is focussed, the `window_title` attribute is set to `""`. You can verify this with:

```bash
leftwm-state | jq '.window_title'
```

For example, let's consider two workspaces: one running a terminal with the  `leftwm-state` command, and another workspace that is completely empty. Moving the cursor to the empty workspace, sets the `window_title` attribute to `""`. However, positioning the cursor on empty space such as padding between windows or the screen border on the non-empty workspace, does not update the `window_title` attribute.

Maybe additional fields should be added to `leftwm-state` to properly determine empty workspaces. Indeed, a `windows` field for each tag would be very useful.

Despite the fact that the `window_title` attribute does not consistently reflect the changes, it could be used as a workaround for identifying when an empty workspace is focussed.

Sadly, I cannot think of a simple solution for detecting which workspace is empty. Specifically, I want to detect which workspace transitions from empty to non-empty, and vice versa.

**Reply by VuiMuich** [reply]
Oops you are right, sorry for the confusion.
As a side effect I found a bug though: when you focus an  empty tag the `window_title` field should be `""` but it still has the value of the last focused window.

And yes exposing more fields in the `leftwm-state` output is definitely considerable. I would suggest a `window_count` field on each tag, this can even be helpful with layouts like `Monocle` or `MainAndDeck` to see how many windows are in the background..

**Reply by trueNAHO** [reply]
I have seen there are several open issues that would be resolved by adding more fields to `leftwm-state`. Does that mean that there is currently no real solution for effectively detecting empty workspaces?

**Reply by nemalex** [reply]
> when you focus an empty tag the window_title field should be "" but it still has the value of the last focused window

I don't find this the case
The window title is unconditionally shown in my eww bar, and is "" correctly when I switch to an empty tag

**Reply by VuiMuich** [reply]
Indeed, that might have been a false positive by me.
I tested running `leftwm-state` from a scratchpad and probably missread the output.

**Reply by trueNAHO** [reply]
> > when you focus an empty tag the window_title field should be "" but it still has the value of the last focused window
> 
> I don't find this the case The window title is unconditionally shown in my eww bar, and is "" correctly when I switch to an empty tag

My `window_title` also properly updates as stated before:

> For example, let's consider two workspaces: one running a terminal with the `leftwm-state` command, and another workspace that is completely empty. Moving the cursor to the empty workspace, sets the `window_title` attribute to `""`.

The only bug I found is:

> However, positioning the cursor on empty space such as padding between windows or the screen border on the non-empty workspace, does not update the `window_title` attribute.

**Reply by nemalex** [reply]
Do you use sloppy mouse focus?

**Reply by trueNAHO** [reply]
> Do you use sloppy mouse focus?

I use the default configuration, but commented out the `scratchpad` entry and all entries in `keybind`. The settings related to "sloppy" seem to be:

```ron
    focus_behaviour: Sloppy,
    sloppy_mouse_follows_focus: true,
```

**Reply by nemalex** [reply]
I found that the focus behavior in this mode is a little funky sometimes
I didn't dig into the code why that is, but I'm planning to, at least, make some detailed bug reports
But this is outside this conversation

Regarding your original issue, checking if the window title is "" is the best approach I can think of to detect an empty tag
Here is a code snippet where I conditionally applying a class to a centerbox
```lisp
(centerbox :class "${wmstate.window_title == "" ? "" : "exampleClass"}"
```

**Reply by trueNAHO** [reply]
> checking if the window title is "" is the best approach I can think of to detect an empty tag Here is a code snippet where I conditionally applying a class to a centerbox
> 
> ```lisp
> (centerbox :class "${wmstate.window_title == "" ? "" : "exampleClass"}"
> ```

This should properly detect focussing empty workspaces. However, in my specific use case, I want to open either one of two `eww` windows based on the presence of a window in the corresponding workspace. Consequently, each workspace will contain a dedicated `eww` window. Therefore, I do not think your proposed solution can handle several workspaces.

`eww` provides several approaches for toggling or hiding windows, but the challenge lies in determining when to activate those commands. This is precisely the moment when workspaces transition between being empty and non-empty.

<details>
<summary>If you are curious how both <em>eww</em> windows would look like</summary>

The initial `eww` window serves as a vertical status bar, while the second `eww` window is a rectangular shape positioned at the center of the workspace. The second window contains multiple sub-rectangles designed to display various specialized information. 

The first window would look like: 

![](https://raw.githubusercontent.com/elkowar/dots-of-war/master/eww-bar/.config/eww-bar/showcase.png)

The second window would look like:

![setup by Axarva](https://raw.githubusercontent.com/Axarva/dotfiles-2.0/472983f5013c2e38752dece94165b0bab5cb1b8c/screenshots/center.png)

</details>
